crates/pgt_workspace/src/workspace/server/parser.rs

```diff
@@ -0,0 +1,374 @@
+        // First check if we have any pending sub-statements to process
+        if let Some((id, range, content)) = self.pending_sub_statements.pop() {
+            if self.filter.apply(&id, &range) {
+                return Some(self.mapper.map(self.parser, id, range, &content));
+            }
+            // If the sub-statement doesn't pass the filter, continue to the next item
+            return self.next();
+        }
+
+        // Process the next top-level statement
+        let next_statement = self.statements.next();
+
+        if let Some((root_id, range, content)) = next_statement {
+            // If we should include sub-statements and this statement has an AST
+            let content_owned = content.to_string();
+            if let Ok(ast) = self
+                .parser
+                .ast_db
+                .load_parse(&root_id, &content_owned)
+                .as_ref()
+            {
+                // Check if this is a SQL function definition with a body
+                if let Some(sub_statement) =
+                    self.parser
+                        .sql_fn_db
+                        .get_function_body(&root_id, ast, &content_owned)
+                {
+                    // Add sub-statements to our pending queue
+                    self.pending_sub_statements.push((
+                        root_id.create_child(),
+                        // adjust range to document
+                        sub_statement.range + range.start(),
+                        sub_statement.body.clone(),
+                    ));
+                }
+            }
+
+            // Return the current statement if it passes the filter
+            if self.filter.apply(&root_id, &range) {
+                return Some(self.mapper.map(self.parser, root_id, range, content));
+            }
+
+            // If the current statement doesn't pass the filter, try the next one
+            return self.next();
+        }
+
+        None
+    }
+}
+
+pub struct DefaultMapper;
+impl<'a> StatementMapper<'a> for DefaultMapper {
+    type Output = (StatementId, TextRange, String);
+
+    fn map(
+        &self,
+        _parser: &'a Parser,
+        id: StatementId,
+        range: TextRange,
+        content: &str,
+    ) -> Self::Output {
+        (id, range, content.to_string())
+    }
+}
+
+pub struct ExecuteStatementMapper;
+impl<'a> StatementMapper<'a> for ExecuteStatementMapper {
+    type Output = (
+        StatementId,
+        TextRange,
+        String,
+        Option<pgt_query_ext::NodeEnum>,
+    );
+
+    fn map(
+        &self,
+        parser: &'a Parser,
+        id: StatementId,
+        range: TextRange,
+        content: &str,
+    ) -> Self::Output {
+        let ast_result = parser.ast_db.load_parse(&id, content);
+        let ast_option = match &*ast_result {
+            Ok(node) => Some(node.clone()),
+            Err(_) => None,
+        };
+
+        (id, range, content.to_string(), ast_option)
+    }
+}
+
```

Comment:
every use case simply creates their own Mapper and defines exactly what data they need :)

Reply:
let's do that separately